File: backend-bars/src/reservations/reservations.controller.ts
import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  UseGuards,
  Request,
} from '@nestjs/common';
import { JwtAuthGuard } from '../auth/jwt-auth.guard';
import { ReservationsService } from './reservations.service';
import {
  CreateReservationDto,
  UpdateReservationDto,
} from './dto/reservation.dto';

@Controller('reservations')
@UseGuards(JwtAuthGuard)
export class ReservationsController {
  constructor(private readonly reservationsService: ReservationsService) {}

  // Crear una nueva reserva
  @Post()
  create(@Request() req, @Body() createReservationDto: CreateReservationDto) {
    return this.reservationsService.create(req.user.id, createReservationDto);
  }

  // Obtener mis reservas (cliente)
  @Get('my-reservations')
  findMyReservations(@Request() req) {
    return this.reservationsService.findMyReservations(req.user.id);
  }

  // Obtener reservas de mis bares (owner)
  @Get('owner-reservations')
  findOwnerReservations(@Request() req) {
    return this.reservationsService.findReservationsForOwner(req.user.id);
  }

  // Obtener una reserva específica
  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.reservationsService.findOne(id);
  }

  // Actualizar una reserva
  @Patch(':id')
  update(
    @Param('id') id: string,
    @Request() req,
    @Body() updateReservationDto: UpdateReservationDto,
  ) {
    return this.reservationsService.update(
      id,
      req.user.id,
      req.user.role,
      updateReservationDto,
    );
  }

  // Cancelar una reserva
  @Patch(':id/cancel')
  cancel(@Param('id') id: string, @Request() req) {
    return this.reservationsService.cancel(id, req.user.id, req.user.role);
  }

  // Confirmar una reserva (solo owner)
  @Patch(':id/confirm')
  confirm(@Param('id') id: string, @Request() req) {
    return this.reservationsService.confirm(id, req.user.id);
  }

  // Completar una reserva (solo owner)
  @Patch(':id/complete')
  complete(@Param('id') id: string, @Request() req) {
    return this.reservationsService.complete(id, req.user.id);
  }

  // Eliminar una reserva
  @Delete(':id')
  remove(@Param('id') id: string, @Request() req) {
    return this.reservationsService.remove(id, req.user.id, req.user.role);
  }
}

File: front-bars-flutter/lib/modules/reservations/controllers/reservations_controller.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';
import 'package:front_bars_flutter/modules/reservations/models/reservation_models.dart';
import 'package:front_bars_flutter/modules/reservations/services/reservations_service.dart';

part 'reservations_controller.g.dart';

@riverpod
class ReservationsController extends _$ReservationsController {
  late final ReservationsService _service;

  @override
  ReservationsState build() {
    _service = ref.watch(reservationsServiceProvider);
    return const ReservationsState();
  }

  // Cargar mis reservas (cliente)
  Future<void> loadMyReservations() async {
    state = state.copyWith(isLoading: true, errorMessage: null);

    try {
      final reservations = await _service.getMyReservations();
      state = state.copyWith(
        reservations: reservations,
        isLoading: false,
        errorMessage: null,
      );
    } catch (e) {
      state = state.copyWith(
        isLoading: false,
        errorMessage: e.toString(),
      );
    }
  }

  // Cargar reservas de los bares del owner
  Future<void> loadOwnerReservations() async {
    state = state.copyWith(isLoading: true, errorMessage: null);

    try {
      final reservations = await _service.getOwnerReservations();
      state = state.copyWith(
        reservations: reservations,
        isLoading: false,
        errorMessage: null,
      );
    } catch (e) {
      state = state.copyWith(
        isLoading: false,
        errorMessage: e.toString(),
      );
    }
  }

  // Cargar una reserva específica
  Future<void> loadReservation(String id) async {
    state = state.copyWith(isLoading: true, errorMessage: null);

    try {
      final reservation = await _service.getReservation(id);
      state = state.copyWith(
        selectedReservation: reservation,
        isLoading: false,
        errorMessage: null,
      );
    } catch (e) {
      state = state.copyWith(
        isLoading: false,
        errorMessage: e.toString(),
      );
    }
  }

  // Crear nueva reserva
  Future<bool> createReservation(Map<String, dynamic> data) async {
    state = state.copyWith(isLoading: true, errorMessage: null);

    try {
      final reservation = await _service.createReservation(data);
      state = state.copyWith(
        reservations: [...state.reservations, reservation],
        isLoading: false,
        errorMessage: null,
      );
      return true;
    } catch (e) {
      state = state.copyWith(
        isLoading: false,
        errorMessage: e.toString(),
      );
      return false;
    }
  }

  // Actualizar reserva
  Future<bool> updateReservation(String id, Map<String, dynamic> data) async {
    state = state.copyWith(isLoading: true, errorMessage: null);

    try {
      final updatedReservation = await _service.updateReservation(id, data);
      final updatedList = state.reservations.map((reservation) {
        return reservation.id == id ? updatedReservation : reservation;
      }).toList();

      state = state.copyWith(
        reservations: updatedList,
        selectedReservation: updatedReservation,
        isLoading: false,
        errorMessage: null,
      );
      return true;
    } catch (e) {
      state = state.copyWith(
        isLoading: false,
        errorMessage: e.toString(),
      );
      return false;
    }
  }

  // Cancelar reserva
  Future<bool> cancelReservation(String id) async {
    state = state.copyWith(isLoading: true, errorMessage: null);

    try {
      await _service.cancelReservation(id);
      // Recargar la reserva para obtener el estado actualizado
      await loadReservation(id);
      return true;
    } catch (e) {
      state = state.copyWith(
        isLoading: false,
        errorMessage: e.toString(),
      );
      return false;
    }
  }

  // Confirmar reserva (owner)
  Future<bool> confirmReservation(String id) async {
    state = state.copyWith(isLoading: true, errorMessage: null);

    try {
      final updatedReservation = await _service.confirmReservation(id);
      final updatedList = state.reservations.map((reservation) {
        return reservation.id == id ? updatedReservation : reservation;
      }).toList();

      state = state.copyWith(
        reservations: updatedList,
        selectedReservation: updatedReservation,
        isLoading: false,
        errorMessage: null,
      );
      return true;
    } catch (e) {
      state = state.copyWith(
        isLoading: false,
        errorMessage: e.toString(),
      );
      return false;
    }
  }

  // Completar reserva (owner)
  Future<bool> completeReservation(String id) async {
    state = state.copyWith(isLoading: true, errorMessage: null);

    try {
      final updatedReservation = await _service.completeReservation(id);
      final updatedList = state.reservations.map((reservation) {
        return reservation.id == id ? updatedReservation : reservation;
      }).toList();

      state = state.copyWith(
        reservations: updatedList,
        selectedReservation: updatedReservation,
        isLoading: false,
        errorMessage: null,
      );
      return true;
    } catch (e) {
      state = state.copyWith(
        isLoading: false,
        errorMessage: e.toString(),
      );
      return false;
    }
  }

  // Eliminar reserva
  Future<bool> deleteReservation(String id) async {
    state = state.copyWith(isLoading: true, errorMessage: null);

    try {
      await _service.deleteReservation(id);
      final updatedList = state.reservations
          .where((reservation) => reservation.id != id)
          .toList();

      state = state.copyWith(
        reservations: updatedList,
        isLoading: false,
        errorMessage: null,
      );
      return true;
    } catch (e) {
      state = state.copyWith(
        isLoading: false,
        errorMessage: e.toString(),
      );
      return false;
    }
  }

  // Limpiar error
  void clearError() {
    state = state.copyWith(errorMessage: null);
  }
}
